components/feature/app-links/src/main/java/mozilla/components/feature/app/links/AppLinksFeature.kt

```diff
@@ -47,65 +55,69 @@ class AppLinksFeature(
     private val loadUrlUseCase: SessionUseCases.DefaultLoadUrlUseCase? = null
 ) : LifecycleAwareFeature {
 
-    @Suppress("DEPRECATION")
-    // TODO migrate to browser-state: https://github.com/mozilla-mobile/android-components/issues/8913
-    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
-    internal val observer = object : mozilla.components.browser.session.SelectionAwareSessionObserver(sessionManager) {
-            override fun onLaunchIntentRequest(
-                session: Session,
-                url: String,
-                appIntent: Intent?
-            ) {
-                if (appIntent == null) {
-                    return
-                }
-
-                val doNotOpenApp = {
-                    loadUrlUseCase?.invoke(url, session.id, EngineSession.LoadUrlFlags.none())
-                }
-
-                val loadUrlAction = {
-                    loadUrlUseCase?.invoke(url, session.id, EngineSession.LoadUrlFlags.none())
-                }
-
-                val doOpenApp = {
-                    useCases.openAppLink(
-                        appIntent,
-                        failedToLaunchAction = failedToLaunchAction,
-                        loadUrlAction = loadUrlAction
-                    )
-                }
-
-                if (!session.private || fragmentManager == null) {
-                    doOpenApp()
-                    return
-                }
-
-                val dialog = getOrCreateDialog()
-                dialog.setAppLinkRedirectUrl(url)
-                dialog.onConfirmRedirect = doOpenApp
-                dialog.onCancelRedirect = doNotOpenApp
-
-                if (!isAlreadyADialogCreated()) {
-                    dialog.showNow(fragmentManager, FRAGMENT_TAG)
-                }
-
-                return
-            }
-    }
+    private var scope: CoroutineScope? = null
 
     /**
      * Starts observing app links on the selected session.
      */
     override fun start() {
-        observer.observeIdOrSelected(sessionId)
+        scope = store.flowScoped { flow ->
+            flow.mapNotNull { state -> state.findTabOrCustomTabOrSelectedTab(sessionId) }
+                .ifChanged {
+                    it.content.appIntent
+                }
+                .collect { tab ->
+                    tab.content.appIntent?.let {
+                        handleAppIntent(tab, it.url, it.appIntent)
+                        store.dispatch(ContentAction.ConsumeAppIntentAction(tab.id))
```

Comment:
Do we need to always consume or only in the case where we actually open the app (via the use case?). I'm not sure how the "open in app" from the menu works, does it need this state to exist?

Reply:
@pocmo I think always after we handled it, because otherwise when the user selects a different tab and then this tab again we would prompt them again, or load the url again, or open the app again depending on settings and private mode.

Reply:
> does it need this state to exist?

It doesn't, nothing else reads the intent from the store after, and we pass along pieces as arguments to `handleAppIntent` and to the new use cases.